### PR TITLE
feat(trie): historical & sidechain state root

### LINKED ISSUE #3057: Invalid ancestor re-org test reports wrong invalid transaction
### Describe the bug

The `Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync` hive test currently fails due to an invalid block. However, it is not the correct invalid block.

The test sets up a valid chain (of length 15) using `forkchoiceUpdated` / `getPayload` / `newPayload`. It then creates an invalid chain in the background, with an invalid block. In this case the invalid block is block 8 (`P8'`).

The simulation fails because we report `SYNCING` (as we download blocks one by one) until we eventually report the wrong block as `INVALID`:
```log
INFO (Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync): Secondary Node has correct block
>> (d2d82b78) {"jsonrpc":"2.0","id":47,"method":"engine_newPayloadV1","params":[{"parentHash":"0xfaf8c0fa94e9f2bdef86fb3774a185fe2ad53a35b144eaf6ebccc6ab7d980939","feeRecipient":"0x0000000000000000000000000000000000000000","stateRoot":"0x96dd2f9b51e6e62e5eb0a61989f84b14f9a5b4143d6e44f48a0784448bdb16f7","receiptsRoot":"0x5d721667f3f41c6e8a5e630b2c5f259855d2365513295c4fd0c5d05498ce6489","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","prevRandao":"0xf0c12e098ed6a02bfe6a6f34b9f0fbf6a37b765b7a9d3f817341f88c25973e0a","blockNumber":"0xf","gasLimit":"0x3043ed","gasUsed":"0xa860","timestamp":"0x1243","extraData":"0x01","baseFeePerGas":"0x8545e76","blockHash":"0xb2efad206aa9c68d7f93720bf8ef8c1f1aeb2de6e95ae0a370d1cbd0e8047fe4","transactions":["0x02f86c0709843b9aca008506fc23ac00830124f89400000000000000000000000000000000000003160180c080a0bb476cdbc61448c8f3df3a60b239ece15aecef4b3e0e2ea83bb826e14d014286a056697ae094e913705310289f44afa0503f911444a5e766fc43cfafb3791eb3da"]}]}
<< (d2d82b78) {"jsonrpc":"2.0","result":{"status":"SYNCING","latestValidHash":null,"validationError":null},"id":47}
INFO (Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync): Response from main client: {SYNCING <nil> <nil>}

... repeats

>> (d2d82b78) {"jsonrpc":"2.0","id":53,"method":"engine_newPayloadV1","params":[{"parentHash":"0xfaf8c0fa94e9f2bdef86fb3774a185fe2ad53a35b144eaf6ebccc6ab7d980939","feeRecipient":"0x0000000000000000000000000000000000000000","stateRoot":"0x96dd2f9b51e6e62e5eb0a61989f84b14f9a5b4143d6e44f48a0784448bdb16f7","receiptsRoot":"0x5d721667f3f41c6e8a5e630b2c5f259855d2365513295c4fd0c5d05498ce6489","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","prevRandao":"0xf0c12e098ed6a02bfe6a6f34b9f0fbf6a37b765b7a9d3f817341f88c25973e0a","blockNumber":"0xf","gasLimit":"0x3043ed","gasUsed":"0xa860","timestamp":"0x1243","extraData":"0x01","baseFeePerGas":"0x8545e76","blockHash":"0xb2efad206aa9c68d7f93720bf8ef8c1f1aeb2de6e95ae0a370d1cbd0e8047fe4","transactions":["0x02f86c0709843b9aca008506fc23ac00830124f89400000000000000000000000000000000000003160180c080a0bb476cdbc61448c8f3df3a60b239ece15aecef4b3e0e2ea83bb826e14d014286a056697ae094e913705310289f44afa0503f911444a5e766fc43cfafb3791eb3da"]}]}
<< (d2d82b78) {"jsonrpc":"2.0","result":{"status":"SYNCING","latestValidHash":null,"validationError":null},"id":53}
INFO (Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync): Response from main client: {SYNCING <nil> <nil>}
>> (d2d82b78) {"jsonrpc":"2.0","id":54,"method":"engine_forkchoiceUpdatedV1","params":[{"headBlockHash":"0xb2efad206aa9c68d7f93720bf8ef8c1f1aeb2de6e95ae0a370d1cbd0e8047fe4","safeBlockHash":"0xb2efad206aa9c68d7f93720bf8ef8c1f1aeb2de6e95ae0a370d1cbd0e8047fe4","finalizedBlockHash":"0x0000000000000000000000000000000000000000000000000000000000000000"},null]}
<< (d2d82b78) {"jsonrpc":"2.0","result":{"payloadStatus":{"status":"INVALID","latestValidHash":"0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5","validationError":"links to previously rejected block"},"payloadId":null},"id":54}
INFO (Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync): Response from main client fcu: {INVALID 0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5 0x4005f51cd0}

... repeats as simulation waits

>> (d2d82b78) {"jsonrpc":"2.0","id":71,"method":"engine_newPayloadV1","params":[{"parentHash":"0xfaf8c0fa94e9f2bdef86fb3774a185fe2ad53a35b144eaf6ebccc6ab7d980939","feeRecipient":"0x0000000000000000000000000000000000000000","stateRoot":"0x96dd2f9b51e6e62e5eb0a61989f84b14f9a5b4143d6e44f48a0784448bdb16f7","receiptsRoot":"0x5d721667f3f41c6e8a5e630b2c5f259855d2365513295c4fd0c5d05498ce6489","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","prevRandao":"0xf0c12e098ed6a02bfe6a6f34b9f0fbf6a37b765b7a9d3f817341f88c25973e0a","blockNumber":"0xf","gasLimit":"0x3043ed","gasUsed":"0xa860","timestamp":"0x1243","extraData":"0x01","baseFeePerGas":"0x8545e76","blockHash":"0xb2efad206aa9c68d7f93720bf8ef8c1f1aeb2de6e95ae0a370d1cbd0e8047fe4","transactions":["0x02f86c0709843b9aca008506fc23ac00830124f89400000000000000000000000000000000000003160180c080a0bb476cdbc61448c8f3df3a60b239ece15aecef4b3e0e2ea83bb826e14d014286a056697ae094e913705310289f44afa0503f911444a5e766fc43cfafb3791eb3da"]}]}
<< (d2d82b78) {"jsonrpc":"2.0","result":{"status":"SYNCING","latestValidHash":null,"validationError":null},"id":71}
INFO (Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync): Response from main client: {SYNCING <nil> <nil>}
>> (d2d82b78) {"jsonrpc":"2.0","id":72,"method":"engine_forkchoiceUpdatedV1","params":[{"headBlockHash":"0xb2efad206aa9c68d7f93720bf8ef8c1f1aeb2de6e95ae0a370d1cbd0e8047fe4","safeBlockHash":"0xb2efad206aa9c68d7f93720bf8ef8c1f1aeb2de6e95ae0a370d1cbd0e8047fe4","finalizedBlockHash":"0x0000000000000000000000000000000000000000000000000000000000000000"},null]}
<< (d2d82b78) {"jsonrpc":"2.0","result":{"payloadStatus":{"status":"INVALID","latestValidHash":"0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5","validationError":"links to previously rejected block"},"payloadId":null},"id":72}
INFO (Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync): Response from main client fcu: {INVALID 0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5 0x400680cb10}
FAIL (Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot, Invalid P8', Reveal using sync): Timeout waiting for main client to detect invalid chain
```

Reth reports that block 6, which the first valid block in the sidechain, and before the invalid block for that sidechain, is invalid:
```log
2023-06-07T05:53:27.014018Z TRACE consensus::engine: start downloading full block. hash=0xce88b40d4b60053eb485e2143bbcd36e87f30d144192b5b8a617bbd1c5f0bde9
2023-06-07T05:53:27.015846Z TRACE consensus::engine: Downloaded full block hash=0xce88b40d4b60053eb485e2143bbcd36e87f30d144192b5b8a617bbd1c5f0bde9 number=8
2023-06-07T05:53:27.015921Z TRACE blockchain_tree: Inserting block hash=0xce88b40d4b60053eb485e2143bbcd36e87f30d144192b5b8a617bbd1c5f0bde9 number=8 parent_hash=0x07cecf1eb7bfee3a485f605e00bcafef25062bb74af3eb192d3296d57dfe3061
2023-06-07T05:53:27.015954Z  INFO try_insert_validated_block{block=(8, 0xce88b40d4b60053eb485e2143bbcd36e87f30d144192b5b8a617bbd1c5f0bde9)}: blockchain_tree: return=Ok(Disconnected { missing_parent: (7, 0x07cecf1eb7bfee3a485f605e00bcafef25062bb74af3eb192d3296d57dfe3061) })
2023-06-07T05:53:27.015961Z TRACE consensus::engine: start downloading full block. hash=0x07cecf1eb7bfee3a485f605e00bcafef25062bb74af3eb192d3296d57dfe3061
2023-06-07T05:53:27.025847Z TRACE consensus::engine: Downloaded full block hash=0x07cecf1eb7bfee3a485f605e00bcafef25062bb74af3eb192d3296d57dfe3061 number=7
2023-06-07T05:53:27.025915Z TRACE blockchain_tree: Inserting block hash=0x07cecf1eb7bfee3a485f605e00bcafef25062bb74af3eb192d3296d57dfe3061 number=7 parent_hash=0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb
2023-06-07T05:53:27.025974Z  INFO try_insert_validated_block{block=(7, 0x07cecf1eb7bfee3a485f605e00bcafef25062bb74af3eb192d3296d57dfe3061)}: blockchain_tree: return=Ok(Disconnected { missing_parent: (6, 0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb) })
2023-06-07T05:53:27.025992Z TRACE consensus::engine: start downloading full block. hash=0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb
2023-06-07T05:53:27.032844Z TRACE consensus::engine: Downloaded full block hash=0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb number=6
2023-06-07T05:53:27.032908Z TRACE blockchain_tree: Inserting block hash=0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb number=6 parent_hash=0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5
2023-06-07T05:53:27.032948Z DEBUG try_insert_validated_block{block=(6, 0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb)}:try_append_canonical_chain: blockchain_tree: Appending block to canonical chain head=0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb parent=(5, 0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5)
2023-06-07T05:53:27.033010Z TRACE try_insert_validated_block{block=(6, 0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb)}:try_append_canonical_chain: blockchain_tree: Executing sidechain block canonical_fork=(5, 0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5)
2023-06-07T05:53:27.033056Z  INFO try_insert_validated_block{block=(6, 0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb)}: blockchain_tree: return=Err(InsertBlockError { error: Execution(Validation(EVM { hash: 0x68578d8e615f14e5ff8babeaa4116e28263f1ddab4d67b7dc3d4502f18c8c6f9, message: "Transaction(LackOfFundForGasLimit { gas_limit: 0x0000000000000000000000000000000000000000000000000007fe5cf2bea001_U256, balance: 0x0000000000000000000000000000000000000000000000000000000000000000_U256 })" })), hash: 0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb, number: 6, parent_hash: 0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5, num_txs: 1, .. })
```
The error is `LackOfFundForGasLimit`:
```
2023-06-07T05:53:27.033069Z DEBUG consensus::engine: Failed to insert downloaded block err=InsertBlockError { error: Execution(Validation(EVM { hash: 0x68578d8e615f14e5ff8babeaa4116e28263f1ddab4d67b7dc3d4502f18c8c6f9, message: "Transaction(LackOfFundForGasLimit { gas_limit: 0x0000000000000000000000000000000000000000000000000007fe5cf2bea001_U256, balance: 0x0000000000000000000000000000000000000000000000000000000000000000_U256 })" })), hash: 0x60e9824348b6e06eeca12b55ed25332a9dc3a3024e64b486a7839755e796cefb, number: 6, parent_hash: 0x5a949798b3aaa04604a28146a9a0ada452b4507e7d9884ad00844249395ca2d5, num_txs: 1, .. }
```

### Steps to reproduce

Run the hive test (it may help to add some traces to `clients/reth/reth.sh`):
```
./hive --client reth --sim "ethereum/engine" --sim.limit "engine-api/Invalid Ancestor Chain Re-Org, Invalid ReceiptsRoot"
```

### Node logs

_No response_

### Platform(s)

_No response_

### What version/commit are you on?

_No response_

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### LINKED ISSUE #5889: Support full payload validations for sidechains
### Describe the feature

Due to past limitations, the state root check was skipped for blocks in sidechains:

https://github.com/paradigmxyz/reth/blob/cd08ba825f63dd60ecb64aa72ae47074e87d24e6/crates/blockchain-tree/src/blockchain_tree.rs#L130-L141

This can cause attestation issues on 1-block reorgs/late blocks specifically with CL clients (lighthouse for example), because ACCEPTED is treated the same as SYNCING
because then the CL needs to wait for processed FCU and can't reuse the previous status response of the payload.

## TODO
* support full state root validation for sidechains in the tree, I believe @Rjected had a PR for this at some point

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Enables sidechain state root computation. Performs complete payload validation even if it does not extend the canonical head.

### CODE REVIEW

**Comment by Rjected** [comment]
currently reviewing, but we should update these docs as well https://github.com/paradigmxyz/reth/blob/92908b190a371a823ce3d4032658badc265294cd/crates/consensus/beacon/src/engine/mod.rs#L1357-L1362

**Comment by rkrasiuk** [comment]
this needs to be fixed as well
https://github.com/paradigmxyz/reth/blob/ce010264b9210bb23d046e1e5f5954096c2887b3/crates/consensus/beacon/src/engine/mod.rs#L1266-L1274

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 1]
```diff
@@ -360,91 +343,64 @@ impl<DB: Database, EF: ExecutorFactory> BlockchainTree<DB, EF> {
...
+
         self.insert_chain(chain);
         self.try_connect_buffered_blocks(block_num_hash);
-        Ok(block_status)
+
+        let status = if block_validation_kind.is_exhaustive() {
+            BlockStatus::Valid
+        } else {
+            BlockStatus::Accepted
+        };
```
Comment (Rjected) [inline crates/blockchain-tree/src/blockchain_tree.rs:402]: makes sense

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 2]
```diff
@@ -456,40 +412,28 @@ impl<DB: Database, EF: ExecutorFactory> BlockchainTree<DB, EF> {
...
-            None => {
-                return Err(InsertBlockError::tree_error(
-                    BlockchainTreeError::BlockSideChainIdConsistency { chain_id: chain_id.into() },
-                    block.block,
-                ))
-            }
-        };
+        let parent_chain = self.state.chains.get_mut(&chain_id).ok_or_else(|| {
+            BlockchainTreeError::BlockSideChainIdConsistency { chain_id: chain_id.into() }
+        })?;
```
Comment (Rjected) [inline crates/blockchain-tree/src/blockchain_tree.rs:430]: yep, this is better

File: crates/trie/src/state.rs [hunk 3]
```diff
@@ -132,6 +132,40 @@ impl HashedPostState {
...
+        );
+        for (hashed_address, account) in other.accounts {
+            accounts.insert(hashed_address, Some(account));
+        }
+        for hashed_address in other.destroyed_accounts {
+            accounts.insert(hashed_address, None);
+        }
+        for (hashed_address, account) in accounts {
+            self.insert_account(hashed_address, account);
+        }
```
Comment (Rjected) [inline crates/trie/src/state.rs:155]: do we have to drain, merge / overwrite, then re-insert here? can we overwrite in-place?

File: crates/trie/src/state.rs [hunk 4]
```diff
@@ -293,13 +328,32 @@ impl HashedStorage {
...
+            ));
+        for (hashed_slot, value) in other.non_zero_valued_slots {
+            entries.insert(hashed_slot, value);
+        }
+        for hashed_slot in other.zero_valued_slots {
+            entries.insert(hashed_slot, U256::ZERO);
+        }
+        for (hashed_slot, value) in entries {
+            self.insert_slot(hashed_slot, value);
+        }
```
Comment (Rjected) [inline crates/trie/src/state.rs:353]: same q w.r.t. draining

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 5]
```diff
@@ -162,15 +161,10 @@ impl<DB: Database, EF: ExecutorFactory> BlockchainTree<DB, EF> {
...
+        if self.is_block_hash_canonical(&block.hash)? || self.is_block_inside_chain(&block) {
             return Ok(Some(BlockStatus::Valid))
         }
 
-        // is block inside chain
-        if let Some(status) = self.is_block_inside_chain(&block) {
-            return Ok(Some(status))
-        }
-
         // check if block is disconnected
```
Comment (Rjected) [inline crates/blockchain-tree/src/blockchain_tree.rs:168]: good simplification

File: crates/blockchain-tree/src/blockchain_tree.rs [hunk 6]
```diff
@@ -724,24 +665,10 @@ impl<DB: Database, EF: ExecutorFactory> BlockchainTree<DB, EF> {
         Ok(())
     }
 
-    /// Check if block is found inside chain and if the chain extends the canonical chain.
-    ///
-    /// if it does extend the canonical chain, return `BlockStatus::Valid`
-    /// if it does not extend the canonical chain, return `BlockStatus::Accepted`
+    /// Returns `true` if block is found inside the tree.
     #[track_caller]
```
Comment (Rjected) [inline crates/blockchain-tree/src/blockchain_tree.rs:669]: can probably remove `track_caller` here because this function shouldn't panic any more

**Review by Rjected — APPROVED** [review]
left some comments and questions, overall lgtm and hive test results look great

File: crates/trie/src/state.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/trie/src/state.rs:155]: for now - yes. i will make a follow-up PR to change how hashed state is represented internally

File: crates/trie/src/state.rs (hunk 4 above)
Comment (rkrasiuk) [reply crates/trie/src/state.rs:353]: same answer as above, not happy about, but it deserves a separate PR

File: crates/trie/src/state.rs (hunk 3 above)
Comment (Rjected) [reply crates/trie/src/state.rs:155]: ok, makes sense for now then

File: crates/interfaces/src/blockchain_tree/mod.rs [hunk 7]
```diff
@@ -182,24 +182,48 @@ impl CanonicalOutcome {
...
+    Canonical,
+    /// The block can be traced back to an ancestor of the canonical head: a historical block, but
+    /// this chain does __not__ include the canonical head.
+    HistoricalFork,
+}
+
+impl BlockAttachment {
+    /// Returns `true` if the block is canonical or a descendant of the canonical head.
+    #[inline]
+    pub fn is_canonical(&self) -> bool {
```
Comment (mattsse) [inline crates/interfaces/src/blockchain_tree/mod.rs:222]:
```suggestion
    pub const fn is_canonical(&self) -> bool {
```

File: crates/blockchain-tree/src/chain.rs [hunk 8]
```diff
@@ -226,45 +192,28 @@ impl AppendableChain {
...
-        parent_block: &SealedHeader,
-        bundle_state_data_provider: BSDP,
-        externals: &TreeExternals<DB, EF>,
-    ) -> RethResult<BundleStateWithReceipts>
-    where
-        BSDP: BundleStateDataProvider,
-        DB: Database,
-        EF: ExecutorFactory,
-    {
-        let (state, _) = Self::validate_and_execute(
```
Comment (mattsse) [inline crates/blockchain-tree/src/chain.rs:257]: good riddance

**Review by mattsse — APPROVED** [review]
the tree and beacon changes lgtm,
can't really speak to the trie changes, but I could follow along.

**Pull request merged**